### ISSUE TITLE
Non-generic object type used in mapped type in true arm of conditional checked for `extends object` can't be validly assigned from an object literal of that same type

### ISSUE THREAD

**Issue by weidehai** [microsoft/TypeScript#50790]
# Bug Report

### 🔎 Search Terms
generic, class extends, keyof class

### 🕗 Version & Regression Information

- This is a crash
- This changed between versions ___4.7.4___ and ___4.7.4___
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about ____10____
- I was unable to test this on prior versions because ___it no need to test  on prior versions___

### ⏯ Playground Link

click it
[Playground link with relevant code](https://www.typescriptlang.org/play?#code/MYGwhgzhAEZtBvAsAKGu6AHATgSwG5gAuAptAPoAWArmQLzQAMqGWehpFEx12xuAewB20BgEZGzNBhwFiZcoRC1R0CVNayOCsCEyV44ySxnUARiFzBoJIWAskAgnoOqAZrogkT6TOcvWbgLYALbEqgDklCQAHhE+WP5W0EoqDBHx0r5J1l4gJMCkACYA-ABc0GYCAvlgQqgAvqhEAJ6YZIaw8LGkQkUwAmYAVgVE0CWI0ADaAArQuCIA1iQtAm5dALplcLMb0A3QFQBMzW1kZgA8ACoAfKpXNjG9-dCDI4Xjk7PzSytr0FctlddvtDtATihUPkxvYytYGMgsjY7A5nPowBUPCAvAAaBJBULECoZPFI1IkYmZJqQlDQ2DAMqXOB3BEJWz2fJogyYzwkUmsAlhIiU-kYcmUxpAA)

### 💻 Code

```ts
class aa {
    private _hue = 0
    private _saturation = 100
    private _value = 100
    private _alpha = 100
    public enableAlpha = false
    public format = 'hex'
    public value = ''
    public selected?: boolean
}
type a = aa extends object ? { [P in keyof aa]:aa[P] } : 2
type b<T> = T extends object ? { [P in keyof T]:T[P] } : 2

let ab:c = {
    enableAlpha: false,
    format: '',
    value: ''
}

let ac:b<aa> = {
    enableAlpha: false,
    format: '',
    value: ''
}
```

### 🙁 Actual behavior

the variable ac type check is good, but ab type check error

### 🙂 Expected behavior
variable ab type check good

**Comment by MartinJohns** [comment]
You forgot to fill out the issue template.

**Comment by whzx5byb** [comment]
repro:
```ts repro
type A = { foo: number }
type B = A extends (A extends object ? {[P in keyof A]: A[P]} : never) ? true : false;
//  ^?
//  3.8.3: type B = true
//  3.9.7: type B = A extends { [P in keyof A]: A[P]; } ? true : false
```

**Comment by weidehai** [comment]
> repro:
> 
> ```ts
> type A = { foo: number }
> type B = A extends (A extends object ? {[P in keyof A]: A[P]} : never) ? true : false;
> //  ^?
> //  3.8.3: type B = true
> //  3.9.7: type B = A extends { [P in keyof A]: A[P]; } ? true : false
> ```
> 
> Seems like `A extends object` has deferred the resolving of `keyof A` unexpectedly when A is not a generic parameter.


Thank you for your answer，as you said,it deferred to resolving type,so when will it be resolved?or it will not be resolved at all？

**Comment by weidehai** [comment]
@whzx5byb

**Comment by weidehai** [comment]
> You forgot to fill out the issue template.

so sorry for that🙇‍♂️，I will continue to complete it，and I hope to get your help❤

**Comment by weidehai** [comment]
@MartinJohns 
i complate to fill out the issue template, if any else need to be added,please tell me

**Comment by weidehai** [comment]
@MartinJohns 
i complate to fill out the issue template, if any else need to be added,please tell me

**Comment by RyanCavanaugh** [comment]
```ts
interface SomeType {
    prop: boolean;
}
type X = SomeType extends object ? { [P in keyof SomeType]: SomeType[P] } : never;

// Error for no obvious reason...
let ab: X = {
    prop: false,
}
```

**Comment by RyanCavanaugh** [comment]
If you could share why you wrote the type this way, that'd be useful. As-is this is not going to be prioritized since there's no reason to write `T extends U ?` when `T` isn't a type parameter and doesn't reference one.

**Comment by weidehai** [comment]
> If you could share why you wrote the type this way, that'd be useful. As-is this is not going to be prioritized since there's no reason to write `T extends U ?` when `T` isn't a type parameter and doesn't reference one.

@RyanCavanaugh I just found this problem by accident and wondered why

**Comment by Tobeyyyyy** [comment]
Seems like this was fixed in 4.9

[Playground](https://www.typescriptlang.org/play?ts=4.9.4#code/JYOwLgpgTgZghgYwgAgMoHsC2EAqBPABxQG8BYAKGSuQKnQIC5kAjddAGwjhAG4KBfCmEIoAGsgC8aLLhHIIAD0ggAJgGdk6ZgCsICMMgD8yYsgDaABWShkAawh50MadnxEAukwyuRl98n5kJhAIADdoPnIKAHpo5AA5dGQAUSg6KGRuPEx0KAgKTgM4ZiZxKTJKalp6Jnh2NQgAGgEKCgR2ODUNODgTCmoaKGBQuEhkAH0ACwBXFCkABn6qoZGx8bVR6ahR4HQQSWQARnnFyqpaYdGUcZH2WYPj04GL1eu4dgJJ3qlHpfPp5jsYAIeQgYqcACCHy+BzqDT+NABQJBMFymFGBwA5JNFJiEQQkcDkLd7lJMXizojAUSGpx9BAVIYmKwOFwQC1yMIiJkDj15EoIKoNFpdPojCZzFYbPZHM4ep4en4AkFkAAmIRyZgAHhwAD4Djh+cp1JodHoDMZTJZrPsZU5kDhPDglYEmOqouRYgkkql0pkQNlcvlyIVMsxVUxvn1KYLwRAoZ84LV3g1mpTUVB0WAmOS0wMSRAcxTBB7Q4gGNqevrygjY4D49Ck8g4U0ERms0W89QC0WBEA)
